### PR TITLE
feat(levm): add contract creation

### PR DESCRIPTION
**Pull request by maximopalopoli**
**Motivation**

Creation transaction is not supported, and it's essential to execute the EVM.

**Description**

Some functionality has been taken from `create()` method of VM. Maybe a refactor can be done in that sense.
In case of error I used the `VMerror` enum, but a future implementation may use other structure

### CODE REVIEW

File: crates/vm/levm/src/errors.rs [hunk 1]
```diff
@@ -26,6 +26,9 @@ pub enum VMError {
     SenderAccountShouldNotHaveBytecode,
     SenderBalanceShouldContainTransferValue,
     GasPriceIsLowerThanBaseFee,
+    AddressAlreadyOccuped,
```
Comment (ilitteri) [inline crates/vm/levm/src/errors.rs:29]:
```suggestion
    AddressAlreadyOccupied,
```

File: crates/vm/levm/src/vm.rs [hunk 2]
```diff
@@ -200,11 +200,202 @@ pub fn word_to_address(word: U256) -> Address {
...
+    (4) Transfer the initial Ether endowment from caller to the new contract
+    (5) Set input data as contract’s deploy code, then execute it with EVM. The ret variable is the returned contract code
+    (6) Check for error. Or if the contract code is too big, fail. Charge the user gas then set the contract code
+    Source: https://medium.com/@hayeah/diving-into-the-ethereum-vm-part-5-the-smart-contract-creation-process-cb7b6133b855
+     */
+    let mut db_copy = db.clone();
+    let mut sender_account = match db_copy.accounts.get(&sender) {
+        Some(acc) => acc,
+        None => {
+            return Err(VMError::SenderAccountDoesNotExist);
```
Comment (ilitteri) [inline crates/vm/levm/src/vm.rs:296]: Although we use a database to track account information, if an account is not in our database, it doesn't mean it doesn't exist but that it has no balance. Accounts are not created in Ethereum; all accounts exist; the only difference is whether they have a balance or not.
```suggestion
            return Err(VMError::OutOfGas);
```
As an exercise, sample a random account and try to send any transaction, or ask its balance.

File: crates/vm/levm/src/vm.rs [hunk 3]
```diff
@@ -200,11 +200,202 @@ pub fn word_to_address(word: U256) -> Address {
...
+
+    // (2)
+    let new_contract_address = match salt {
+        Some(salt) => VM::calculate_create2_address(sender, &calldata, salt),
+        None => VM::calculate_create_address(sender, sender_account.nonce),
+    };
+
+    // If address is already in db, there's an error
+    if db_copy.accounts.contains_key(&new_contract_address) {
+        return Err(VMError::AddressAlreadyOccuped);
```
Comment (ilitteri) [inline crates/vm/levm/src/vm.rs:319]:
```suggestion
        return Err(VMError::AddressAlreadyOccupied);
```

File: crates/vm/levm/src/vm.rs [hunk 4]
```diff
@@ -200,11 +200,202 @@ pub fn word_to_address(word: U256) -> Address {
...
+
+    // (1)
+    if sender_account.balance < value {
+        return Err(VMError::OutOfGas); // Maybe a more personalized error
+    }
+
+    db_copy.increment_account_nonce(&sender);
+    // Check for nonce errors?
+
+    sender_account.balance -= value;
```
Comment (ilitteri) [inline crates/vm/levm/src/vm.rs:309]: Is it possible that this is already done in (4)?

File: crates/vm/levm/src/vm.rs [hunk 5]
```diff
@@ -200,11 +200,202 @@ pub fn word_to_address(word: U256) -> Address {
...
+        None,
+    )?;
+
+    let res = vm.transact()?;
+    // Don't use a revert bc work with clones, so don't have to save previous state
+
+    let contract_code = res.output;
+
+    // (6)
+    if contract_code.len() > 24576 {
```
Comment (ilitteri) [inline crates/vm/levm/src/vm.rs:367]: Wrap `24576` into a constant

File: crates/vm/levm/src/vm.rs [hunk 6]
```diff
@@ -200,11 +200,202 @@ pub fn word_to_address(word: U256) -> Address {
...
+        return Err(VMError::ContractOutputTooBig);
+    }
+    // Supposing contract code has contents
+    if contract_code[0] == 0xef {
```
Comment (ilitteri) [inline crates/vm/levm/src/vm.rs:371]: Use a constant for `0xef`.

File: crates/vm/levm/src/vm.rs [hunk 7]
```diff
@@ -200,11 +200,202 @@ pub fn word_to_address(word: U256) -> Address {
...
+    }
+
+    // If the initialization code completes successfully, a final contract-creation cost is paid,
+    // the code-deposit cost, c, proportional to the size of the created contract’s code
+    let creation_cost = 200 * contract_code.len();
+    if creation_cost > sender_account.balance.as_usize() {
+        return Err(VMError::OutOfGas);
+    }
+
+    sender_account.balance -= U256::from(creation_cost);
```
Comment (ilitteri) [inline crates/vm/levm/src/vm.rs:382]: Nit.
```suggestion
    sender_account.balance.checked_sub(U256::from(creation_cost)).ok_or(VMError::OutOfGas)?;
```

**Review by ilitteri — CHANGES_REQUESTED** [review]
The PR is looking good. Left some nits and suggestions.

File: crates/vm/levm/src/vm.rs (hunk 6 above)
Comment (maximopalopoli) [reply crates/vm/levm/src/vm.rs:371]: I named it INVALID_CONTRACT_PREFIX, it's ok?

**Review by jrchatruc — APPROVED** [review]

File: crates/vm/levm/src/vm.rs (hunk 6 above)
Comment (ilitteri) [reply crates/vm/levm/src/vm.rs]: Yes!

File: crates/vm/levm/src/vm.rs [hunk 8]
```diff
@@ -200,11 +200,201 @@ pub fn word_to_address(word: U256) -> Address {
...
+    }
+
+    // If the initialization code completes successfully, a final contract-creation cost is paid,
+    // the code-deposit cost, c, proportional to the size of the created contract’s code
+    let creation_cost = 200 * contract_code.len();
+
+    sender_account
+        .balance
+        .checked_sub(U256::from(creation_cost))
+        .ok_or(VMError::OutOfGas)?;
```
Comment (ilitteri) [inline crates/vm/levm/src/vm.rs:380]: I do not know why this didn't make any test fail, but I have missed something in my previous suggestion
```suggestion
    sender_account = sender_account
        .balance
        .checked_sub(U256::from(creation_cost))
        .ok_or(VMError::OutOfGas)?;
```

Comment (maximopalopoli) [reply crates/vm/levm/src/vm.rs:380]: Wouldn't be
```
    sender_account.balance = sender_account
        .balance
        .checked_sub(U256::from(creation_cost))
        .ok_or(VMError::OutOfGas)?;
```
?

**Review by ilitteri — APPROVED** [review]
LGTM

**Pull request merged**